docs/icicle/primitives/poseidon.md

```diff
@@ -0,0 +1,228 @@
+
+This is done to prevent collions and to prevent certain cryptographic attacks by insuring that the internal state is sufficiently mixed and unpredictable,
+
+![Alt text](image.png)
+
+### Applying full and partial rounds
+
+To generate a secure hash output the algorithm goes through a series of "full round" and "partial rounds" as well as transformations between these sets of rounds.
+
+First full rounds => apply SBox and Round constants => partial rounds => Last full rounds => Apply SBox
+
+#### Full rounds
+
+![Alt text](image-1.png)
+
+**Uniform Application of S-Box:** In full rounds, the S-box (a non-linear transformation) is applied uniformly to every element of the hash function's internal state. This ensures a high degree of mixing and diffusion, contributing to the hash function's security. The S-box involves raising each element of the state to a certain power denoted by `α` is a member of the finite field defined by the prime `p`, `α` can be selected by the user.
+
+**Linear Transformation:** After applying the S-box, a linear transformation is performed on the state. This involves multiplying the state by a MDS (Maximum Distance Separable) Matrix. which further diffuses the transformations applied by the S-box across the entire state.
+
+**Addition of Round Constants:** Each element of the state is then modified by adding a unique round constant. These constants are different for each round and are precomputed as part of the hash function's initialization. The addition of round constants ensures that even minor changes to the input produce significant differences in the output.
+
+#### Partial Rounds
+
+**Selective Application of S-Box:** Partial rounds apply the S-box transformation to only one element of the internal state per round, rather than to all elements. This selective application significantly reduces the computational complexity of the hash function without compromising its security. The choice of which element to apply the S-box to can follow a specific pattern or be fixed, depending on the design of the hash function.
+
+**Linear Transformation and Round Constants:** A linear transformation is performed, and round constants are added. However, the linear transformation in partial rounds can be designed to be less computationally intensive (this is done by using a sparse matrix) than in full rounds, further optimizing the function's efficiency.
+
+
+The user of a Poseidon hash often can choose how many partial or full rounds he wishes to apply, more full rounds will increase security and degrade performance. The choice and balance depend highly on the use case.
+
+![Alt text](image-2.png)
+
+## Using Poseidon Hash
+
+ICICLE Poseidon hash is implemented for GPU and parallelization is performed for each element of the state rather than for each state.
```

Comment:
It's important to note that our implementation is focused on calculating multiple hash-sums over multiple pre-images in parallel. Rather than going block by block over the input vector.

So for Poseidon of arity 2 and input of size 1024 * 2, you will get 1024 elements of output. Which is the result of calculating 1024 Poseidon hashes over blocks of size 2.

Reply:
Good point, I copied ur explanation into the docs sine i think its simple and clear.